ci-automation: improve bincache garbage collection, cover cached release artifacts

t-lo opened the pull request:
This PR extends the garbage collector for the build cache server to remove cached release artifacts. Release artifacts are copied to the official mirrors and do not need to remain on the build cache after a release was published.

By default, the 10 latest releases per channel (including LTS and previous LTS) are kept.

Also excluded from garbage collection are:
 - Emerging new major releases (i.e. major number larger than the latest Alpha release)
 - channel progressions (major number exists in the lists of releases to keep but minor is bigger than any release)
 - patch releases (major and minor exist in list of releases to keep but patch level is newer than in any release)
 - SDKs (tarballs and containers) of any release in the list of releases to keep; i.e. the SDK in <MAJOR>.0.0 for any release to keep.

Furthermore, this change unsets `POSIXLY_CORRECT` for all `awk` calls that use `gensub`. This fixes an issue with restricted environments like e.g. the systemd-run env we use in our ci-automation.

# How to use
```bash
bash -c "source ci-automation/garbage_collect_releases.sh ; DRY_RUN=y garbage_collect_releases;"
```

# Testing done
Ran the above command and captured the output:  [remove_releases.txt](https://github.com/flatcar/scripts/files/14099392/remove_releases.txt). Reviewed the output; looks sane.

krnowak commented on ci-automation/garbage_collect_releases.sh:
```diff
@@ -0,0 +1,180 @@
+                    minor = gensub("[0-9]+\\.([0-9]+)\\.[0-9]+","\\1","g") + 0
+                    patch = gensub("[0-9]+\\.[0-9]+\\.([0-9]+)","\\1","g") + 0
+
+                    if (   ((path == "sdk/amd64") || (path == "containers")) \
+                        && (vmajor == major) && (vminor == 0) && (vpatch == 0) ) {
+                        print ""
+                        print "## Skipping " version " in " path " because it contains the SDK for release " $0 " in keep list."
+                        ret = 0
+                        exit
+                    }
```
That's tricky. We seldom had a case where we had to rebuild SDK for some patch release, so SDK ended up having a version different than `<MAJOR>.0.0`.

t-lo replied:
Yeah, I agree, the check is too rigorous. We should skip all versions that have a path below `/sdk/smd64` and that match a major version on the keep list.

t-lo commented:
(I'm not around for more than a week, feel free to amend and merge this PR as you see fit)

t-lo commented on ci-automation/garbage_collect_releases.sh:
```diff
@@ -0,0 +1,180 @@
+    local sshcmd="$(gen_sshcmd)"
+
+    local keep_versions
+    mapfile -t keep_versions < <(unset POSIXLY_CORRECT; \
```
Good catch, I'll rewrite.

t-lo commented:
Addressed all comments, ran another dry run using
```bash
bash -c "source ci-automation/garbage_collect_releases.sh ; DRY_RUN=y garbage_collect_releases;"
```

Output: 
[out.txt](https://github.com/flatcar/scripts/files/14241513/out.txt)

krnowak commented on ci-automation/garbage_collect_releases.sh:
```diff
@@ -0,0 +1,183 @@
+    source ci-automation/ci_automation_common.sh
+    local sshcmd="$(gen_sshcmd)"
+
+    local keep="$( 
```
Quotes in assignment are not necessary.
```suggestion
    local keep=$( 
```

krnowak commented on ci-automation/garbage_collect_releases.sh:
```diff
@@ -0,0 +1,183 @@
+                        chan_count["lts"][major] = chan_count["lts"][major] + 1
+                    }
+                } '
+    )"
```
```suggestion
    )
```

krnowak commented on ci-automation/garbage_collect_releases.sh:
```diff
@@ -0,0 +1,183 @@
+
+            # skip if version is marked for keeping OR if it's a new release about to be published
+            if printf "%s\n" "${keep_versions[@]}" \
+               | { unset POSIXLY_CORRECT ; awk -v version="${version}" -v path="${dir}" '
```
Can I ask you to rename some of the variables? It would make it easier to track which version, major, minor and patch is which. What I mean is something like renaming `version` to `bincache_version`, `v{major,minor,patch}` to `bincache_{major,minor,patch}`. And then possibly having an assignment like `keep_version = $0` and variables `keep_{major,minor,patch}`. `major_alpha` probably could be `keep_major_alpha`.

t-lo replied:
Yes, good point. Tbh I got confused myself when updating the PR after one week of vacation :sweat_smile:

t-lo commented:
Addressed all comments, ran another dry run using
```bash

bash -c "source ci-automation/garbage_collect_releases.sh ; DRY_RUN=y garbage_collect_releases;"
```

Output: [out.txt](https://github.com/flatcar/scripts/files/14284257/out.txt)
